core/src/loader.rs

```diff
@@ -1512,26 +1480,63 @@ impl<'gc> Loader<'gc> {
                         // Flash always fires an initial 'progress' event with
                         // bytesLoaded=0 and bytesTotal set to the proper value.
                         // This only seems to happen for an AVM2 event handler
-                        Loader::movie_loader_progress(handle, uc, 0, length)?;
+                        Loader::movie_loader_progress(handle, &mut activation.context, 0, length)?;
                     }
 
-                    let movie = Arc::new(SwfMovie::from_loaded_image(url, length));
+                    let library = activation.context.library.library_for_movie_mut(movie);
 
+                    if let Some(domain) = domain {
+                        library.set_avm2_domain(domain);
+                    }
+
+                    // This will construct AVM2-side objects even under AVM1, but it doesn't matter,
+                    // since Bitmap and BitmapData never have AVM1-side objects.
                     let bitmap = ruffle_render::utils::decode_define_bits_jpeg(data, None)?;
-                    let bitmap_obj = Bitmap::new(uc, 0, bitmap)?;
 
-                    if let Some(mut mc) = clip.as_movie_clip() {
-                        let mut activation = Avm2Activation::from_nothing(uc.reborrow());
-                        mc.replace_with_movie(&mut activation.context, Some(movie), None);
-                        mc.replace_at_depth(uc, bitmap_obj.into(), 1);
```

Comment:
I do not understand or know this part of the codebase, but, further down in the changes, `replace_at_depth()` is apparently called with AVM1. Should it also be called with AVM2 somewhere?

Reply:
No - under AVM2, we manually add it as a child of the `Loader` instance.